File: backend-node/src/routes/crypto.js
import express from 'express';
import { pool } from '../db.js';
import { cryptoUsdRate, validCryptoAmount } from '../utils.js';

const router = express.Router();
const lastEarnByUser = new Map();

router.get('/rate', (_req, res) => {
  return res.json({ symbol: 'BTC/USD', rate: cryptoUsdRate(), updated_every_seconds: 10 });
});

router.post('/earn', async (req, res, next) => {
  const now = Date.now();
  const last = lastEarnByUser.get(req.user.id) || 0;
  const remaining = 60000 - (now - last);
  if (remaining > 0) return res.status(429).json({ error: 'Passive income can be collected once per minute', retry_after_seconds: Math.ceil(remaining / 1000) });
  try {
    const result = await pool.query(
      `UPDATE users SET crypto_balance = crypto_balance + 0.000010
       WHERE id = $1
       RETURNING crypto_balance`,
      [req.user.id]
    );
    if (!result.rows[0]) return res.status(404).json({ error: 'User not found' });
    lastEarnByUser.set(req.user.id, now);
    return res.json({ success: true, earned_crypto: '0.00001000', crypto_balance: result.rows[0].crypto_balance, rate: cryptoUsdRate() });
  } catch (error) {
    return next(error);
  }
});

router.post('/sell', async (req, res, next) => {
  const cardId = String(req.body?.card_id || '');
  const amountCrypto = String(req.body?.amount_crypto ?? '');
  if (!/^[0-9a-fA-F-]{36}$/.test(cardId)) return res.status(400).json({ error: 'Valid card_id is required' });
  if (!validCryptoAmount(amountCrypto)) return res.status(400).json({ error: 'Crypto amount must be positive with up to 8 decimals' });
  const rate = cryptoUsdRate();
  const client = await pool.connect();
  try {
    await client.query('BEGIN');
    const userResult = await client.query(
      `SELECT crypto_balance FROM users WHERE id = $1 FOR UPDATE`,
      [req.user.id]
    );
    if (!userResult.rows[0]) {
      await client.query('ROLLBACK');
      return res.status(404).json({ error: 'User not found' });
    }
    const userCrypto = userResult.rows[0].crypto_balance;
    const check = await client.query(`SELECT ($1::numeric >= $2::numeric) AS enough`, [userCrypto, amountCrypto]);
    if (!check.rows[0].enough) {
      await client.query('ROLLBACK');
      return res.status(400).json({ error: 'Insufficient crypto balance' });
    }
    const cardResult = await client.query(
      `SELECT id FROM cards WHERE id = $1 AND user_id = $2 FOR UPDATE`,
      [cardId, req.user.id]
    );
    if (!cardResult.rows[0]) {
      await client.query('ROLLBACK');
      return res.status(404).json({ error: 'Destination card not found' });
    }
    const userUpdated = await client.query(
      `UPDATE users SET crypto_balance = crypto_balance - $1::numeric
       WHERE id = $2 RETURNING crypto_balance`,
      [amountCrypto, req.user.id]
    );
    const cardUpdated = await client.query(
      `UPDATE cards SET balance = balance + ($1::numeric * $2::numeric)
       WHERE id = $3 RETURNING balance`,
      [amountCrypto, rate, cardId]
    );
    const usdAmount = await client.query(`SELECT ($1::numeric * $2::numeric)::numeric AS value`, [amountCrypto, rate]);
    await client.query(
      `INSERT INTO transactions (sender_card_id, receiver_card_id, amount, type)
       VALUES (NULL, $1, $2::numeric, 'crypto_exchange')`,
      [cardId, usdAmount.rows[0].value]
    );
    await client.query('COMMIT');
    return res.json({ success: true, rate, usd_amount: usdAmount.rows[0].value, crypto_balance: userUpdated.rows[0].crypto_balance, card_balance: cardUpdated.rows[0].balance });
  } catch (error) {
    await client.query('ROLLBACK');
    return next(error);
  } finally {
    client.release();
  }
});

export default router;

File: backend-python/app/routes/crypto.py
import uuid
import time
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session
from ..auth import get_current_user
from ..database import get_db
from ..models import Card, Transaction, User
from ..schemas import CryptoSellBody
from ..utils import crypto_usd_rate, valid_crypto_amount

router = APIRouter(prefix='/api/crypto', tags=['crypto'])
last_earn = {}

@router.get('/rate')
def rate(user: User = Depends(get_current_user)):
    return {'symbol': 'BTC/USD', 'rate': crypto_usd_rate(), 'updated_every_seconds': 10}

@router.post('/earn')
def earn(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    now = time.monotonic()
    last = last_earn.get(user.id, 0)
    remaining = 60 - (now - last)
    if remaining > 0:
        raise HTTPException(429, detail={'message': 'Passive income can be collected once per minute', 'retry_after_seconds': int(remaining) + 1})
    with db.begin():
        current = db.scalar(select(User).where(User.id == user.id).with_for_update())
        if not current:
            raise HTTPException(404, 'User not found')
        current.crypto_balance = Decimal(current.crypto_balance) + Decimal('0.000010')
        db.flush()
        balance = current.crypto_balance
    last_earn[user.id] = now
    return {'success': True, 'earned_crypto': '0.00001000', 'crypto_balance': str(balance), 'rate': crypto_usd_rate()}

@router.post('/sell')
def sell(body: CryptoSellBody, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    try:
        card_id = uuid.UUID(body.card_id)
    except ValueError:
        raise HTTPException(400, 'Valid card_id is required')
    if not valid_crypto_amount(body.amount_crypto):
        raise HTTPException(400, 'Crypto amount must be positive with up to 8 decimals')
    amount_crypto = Decimal(body.amount_crypto)
    rate_now = Decimal(str(crypto_usd_rate()))
    with db.begin():
        current = db.scalar(select(User).where(User.id == user.id).with_for_update())
        if not current:
            raise HTTPException(404, 'User not found')
        if Decimal(current.crypto_balance) < amount_crypto:
            raise HTTPException(400, 'Insufficient crypto balance')
        card = db.scalar(select(Card).where(Card.id == card_id, Card.user_id == user.id).with_for_update())
        if not card:
            raise HTTPException(404, 'Destination card not found')
        usd_amount = amount_crypto * rate_now
        current.crypto_balance = Decimal(current.crypto_balance) - amount_crypto
        card.balance = Decimal(card.balance) + usd_amount
        db.add(Transaction(receiver_card_id=card.id, amount=usd_amount, type='crypto_exchange'))
        db.flush()
        crypto_balance = current.crypto_balance
        card_balance = card.balance
    return {'success': True, 'rate': int(rate_now), 'usd_amount': str(usd_amount), 'crypto_balance': str(crypto_balance), 'card_balance': str(card_balance)}
